File: tyhp-lang/phpstorm/src/main/kotlin/com/tyhp/lang/workspace/PathUtils.kt
package com.tyhp.lang.workspace

/** POSIX-style path helpers so membership matches `tyhp build` on every OS. */

fun toPosix(path: String): String = path.replace('\\', '/')

fun stripDotSlash(path: String): String {
    var out = toPosix(path).trim()
    while (out.startsWith("./")) {
        out = out.substring(2)
    }
    return out
}

fun posixNormalize(path: String): String {
    val posix = toPosix(path)
    val absolute = posix.startsWith("/")
    val parts = posix.split('/').filter { it.isNotEmpty() && it != "." }
    val stack = ArrayDeque<String>()
    for (part in parts) {
        if (part == "..") {
            if (stack.isNotEmpty() && stack.last() != "..") {
                stack.removeLast()
            } else if (!absolute) {
                stack.addLast("..")
            }
        } else {
            stack.addLast(part)
        }
    }
    val joined = stack.joinToString("/")
    return if (absolute) "/$joined" else joined.ifEmpty { "." }
}

fun posixDirname(path: String): String {
    val n = toPosix(path).trimEnd('/')
    val i = n.lastIndexOf('/')
    return when {
        i < 0 -> "."
        i == 0 -> "/"
        else -> n.substring(0, i)
    }
}

fun posixBasename(path: String): String {
    val n = toPosix(path).trimEnd('/')
    val i = n.lastIndexOf('/')
    return if (i < 0) n else n.substring(i + 1)
}

private fun splitAbs(path: String): List<String> {
    val n = posixNormalize(path)
    if (n == "/") {
        return emptyList()
    }
    return n.trimStart('/').split('/').filter { it.isNotEmpty() }
}

fun posixRelative(fromDir: String, toPath: String, caseInsensitive: Boolean = false): String {
    val from = splitAbs(fromDir)
    val to = splitAbs(toPath)
    var i = 0
    val n = minOf(from.size, to.size)
    while (i < n) {
        val left = if (caseInsensitive) from[i].lowercase() else from[i]
        val right = if (caseInsensitive) to[i].lowercase() else to[i]
        if (left != right) {
            break
        }
        i += 1
    }
    val ups = List(from.size - i) { ".." }
    val down = to.drop(i)
    val parts = ups + down
    return parts.joinToString("/").ifEmpty { "." }
}

fun pathHops(fromDir: String, toPath: String, caseInsensitive: Boolean = false): Int {
    val rel = posixRelative(fromDir, toPath, caseInsensitive)
    if (rel == "." || rel.isEmpty()) {
        return 0
    }
    return rel.split('/').count { it.isNotEmpty() }
}

fun isPathInside(dir: String, filePath: String, caseInsensitive: Boolean = false): Boolean {
    val rel = posixRelative(dir, filePath, caseInsensitive)
    return rel != ".." && !rel.startsWith("../") && rel != "."
}

fun hasAncestorTyhpJson(
    filePath: String,
    workspaceRoot: String?,
    exists: (String) -> Boolean,
    join: (String, String) -> String = { dir, name -> "${toPosix(dir).trimEnd('/')}/$name" },
): Boolean {
    if (workspaceRoot.isNullOrBlank()) {
        return false
    }
    val stop = posixNormalize(workspaceRoot)
    var dir = posixDirname(filePath)
    val seen = HashSet<String>()
    while (seen.add(dir)) {
        if (exists(join(dir, "tyhp.json"))) {
            return true
        }
        val normalized = posixNormalize(dir)
        if (normalized == stop) {
            break
        }
        if (stop != "/" && !isPathInside(stop, dir) && normalized != stop) {
            break
        }
        val parent = posixDirname(dir)
        if (parent == dir) {
            break
        }
        dir = parent
    }
    return false
}

val INDEX_SKIP_DIR_NAMES: Set<String> = setOf("node_modules", "vendor", ".git", "bin", "obj", "dist", "build")

fun shouldSkipIndexedTyhpJson(filePath: String): Boolean =
    toPosix(filePath).split('/').any { it in INDEX_SKIP_DIR_NAMES }

fun matchingWorkspaceRoot(
    filePath: String,
    workspaceRoots: List<String>,
    caseInsensitive: Boolean = false,
): String? {
    val matches = workspaceRoots.filter { root ->
        val n = posixNormalize(root)
        val file = posixNormalize(filePath)
        if (caseInsensitive) {
            file.lowercase() == n.lowercase() || isPathInside(n, file, true)
        } else {
            file == n || isPathInside(n, file, false)
        }
    }
    return matches.maxByOrNull { it.length }
}

File: tyhp-lang/vscode/src/workspace/pathUtils.ts
/**
 * POSIX-style path helpers so membership / owner selection match
 * `tyhp build` (globs relative to the `tyhp.json` directory) on every OS.
 */

export function toPosix(p: string): string {
    return p.replace(/\\/g, "/");
}

export function stripDotSlash(p: string): string {
    let out = toPosix(p).trim();
    while (out.startsWith("./")) {
        out = out.slice(2);
    }
    return out;
}

export function posixNormalize(p: string): string {
    const posix = toPosix(p);
    const absolute = posix.startsWith("/");
    const parts = posix.split("/").filter((part, i) => part !== "" && part !== "." && !(i === 0 && part === ""));
    const stack: string[] = [];
    for (const part of parts) {
        if (part === "..") {
            if (stack.length > 0 && stack[stack.length - 1] !== "..") {
                stack.pop();
            } else if (!absolute) {
                stack.push("..");
            }
        } else {
            stack.push(part);
        }
    }
    const joined = stack.join("/");
    if (absolute) {
        return `/${joined}`;
    }
    return joined === "" ? "." : joined;
}

export function posixDirname(p: string): string {
    const n = toPosix(p).replace(/\/+$/, "");
    const i = n.lastIndexOf("/");
    if (i < 0) {
        return ".";
    }
    if (i === 0) {
        return "/";
    }
    return n.slice(0, i);
}

export function posixBasename(p: string): string {
    const n = toPosix(p).replace(/\/+$/, "");
    const i = n.lastIndexOf("/");
    return i < 0 ? n : n.slice(i + 1);
}

function splitAbs(p: string): string[] {
    const n = posixNormalize(p);
    if (n === "/") {
        return [];
    }
    return n.replace(/^\//, "").split("/").filter((part) => part !== "");
}

function commonPrefixLength(a: readonly string[], b: readonly string[], caseInsensitive: boolean): number {
    const n = Math.min(a.length, b.length);
    let i = 0;
    while (i < n) {
        const left = caseInsensitive ? a[i].toLowerCase() : a[i];
        const right = caseInsensitive ? b[i].toLowerCase() : b[i];
        if (left !== right) {
            break;
        }
        i += 1;
    }
    return i;
}

/**
 * Relative POSIX path from [fromDir] to [toPath] (`../` when the target is
 * outside [fromDir]). Matches Node `path.relative` for absolute POSIX paths.
 */
export function posixRelative(fromDir: string, toPath: string, caseInsensitive = false): string {
    const from = splitAbs(fromDir);
    const to = splitAbs(toPath);
    const i = commonPrefixLength(from, to, caseInsensitive);
    const ups = from.length - i;
    const down = to.slice(i);
    const parts = [...Array.from({ length: ups }, () => ".."), ...down];
    return parts.join("/") || ".";
}

/** Directory distance (path components) from [fromDir] to [toPath]. */
export function pathHops(fromDir: string, toPath: string, caseInsensitive = false): number {
    const rel = posixRelative(fromDir, toPath, caseInsensitive);
    if (rel === "." || rel === "") {
        return 0;
    }
    return rel.split("/").filter((part) => part !== "").length;
}

/**
 * True when [filePath] is the directory [dir] or a descendant of it.
 */
export function isPathInside(dir: string, filePath: string, caseInsensitive = false): boolean {
    const rel = posixRelative(dir, filePath, caseInsensitive);
    return rel !== ".." && !rel.startsWith("../") && rel !== ".";
}

/**
 * Walk from [filePath]'s directory up to [workspaceRoot], returning true if
 * any ancestor directory contains `tyhp.json`.
 */
export function hasAncestorTyhpJson(
    filePath: string,
    workspaceRoot: string | undefined,
    exists: (path: string) => boolean,
    join: (dir: string, name: string) => string = (dir, name) => `${toPosix(dir).replace(/\/$/, "")}/${name}`
): boolean {
    if (!workspaceRoot) {
        return false;
    }
    const stop = posixNormalize(workspaceRoot);
    let dir = posixDirname(filePath);
    const seen = new Set<string>();
    while (!seen.has(dir)) {
        seen.add(dir);
        if (exists(join(dir, "tyhp.json"))) {
            return true;
        }
        const normalized = posixNormalize(dir);
        if (normalized === stop) {
            break;
        }
        if (stop !== "/" && !isPathInside(stop, dir) && normalized !== stop) {
            break;
        }
        const parent = posixDirname(dir);
        if (parent === dir) {
            break;
        }
        dir = parent;
    }
    return false;
}

export const INDEX_SKIP_DIR_NAMES = new Set(["node_modules", "vendor", ".git", "bin", "obj", "dist", "build"]);

/** True when a `tyhp.json` path should be ignored during workspace indexing. */
export function shouldSkipIndexedTyhpJson(filePath: string): boolean {
    const parts = toPosix(filePath).split("/");
    return parts.some((part) => INDEX_SKIP_DIR_NAMES.has(part));
}

export function matchingWorkspaceRoot(
    filePath: string,
    workspaceRoots: readonly string[],
    caseInsensitive = false
): string | undefined {
    const matches = workspaceRoots.filter((root) => {
        const n = posixNormalize(root);
        const file = posixNormalize(filePath);
        if (caseInsensitive) {
            return file.toLowerCase() === n.toLowerCase() || isPathInside(n, file, true);
        }
        return file === n || isPathInside(n, file, false);
    });
    if (matches.length === 0) {
        return undefined;
    }
    return matches.reduce((best, cur) => (cur.length >= best.length ? cur : best));
}
